src/containers/s21_list.h

```diff
@@ -0,0 +1,487 @@
+
+  void clear();
+  iterator insert(iterator pos, const_reference value);
+  void push_back(const_reference data) noexcept;
+  void pop_back();
+  void push_front(const_reference value) noexcept;
+  void pop_front();
+  void merge(list &other);
+  void splice(const_iterator pos, list &other);
+  void sort();
+  void reverse() noexcept;
+  void unique();
+  void erase(iterator pos);
+  const_reference front();
+  const_reference back();
+
+ private:
+  Node *m_head;
+  Node *m_tail;
+  size_type m_size;
+  Node *m_end;
+};
+}  // namespace s21
+
+template <typename T>
+s21::list<T>::ListIterator::ListIterator() noexcept : current_(nullptr) {}
+
+template <typename T>
+s21::list<T>::ListIterator::ListIterator(Node *node) noexcept
+    : current_(node) {}
+
+template <typename T>
+typename s21::list<T>::reference s21::list<T>::ListIterator::operator*() const {
+  return current_->value_;
+};
+
+template <typename T>
+typename s21::list<T>::ListIterator &s21::list<T>::ListIterator::operator++() {
+  current_ = current_->next_;
+  return *this;
+}
+
+template <typename T>
+typename s21::list<T>::ListIterator s21::list<T>::ListIterator::operator++(
+    int) {
+  ListIterator temp(current_);
+  current_ = current_->next_;
+  return temp;
+}
+
+template <typename T>
+typename s21::list<T>::ListIterator &s21::list<T>::ListIterator::operator--() {
+  current_ = current_->prev_;
+  return *this;
+}
+
+template <typename T>
+typename s21::list<T>::ListIterator s21::list<T>::ListIterator::operator--(
+    int) {
+  ListIterator temp(current_);
+  current_ = current_->prev_;
+  return temp;
+}
+
+template <typename T>
+bool s21::list<T>::ListIterator::operator==(
+    const s21::list<T>::ListIterator &other) const {
+  return current_ == other.current_;
+}
+
+template <typename T>
+bool s21::list<T>::ListIterator::operator!=(const ListIterator &other) const {
+  return !(*this == other);
+}
+
+template <typename T>
+typename s21::list<T>::const_reference
+s21::list<T>::ListConstIterator::operator*() const {
+  return this->current_->value_;
+}
+
+template <typename T>
+typename s21::list<T>::iterator s21::list<T>::begin() noexcept {
+  return iterator(m_head ? this->m_head : this->m_end);
+}
+
+template <typename T>
+typename s21::list<T>::const_iterator s21::list<T>::cbegin() const noexcept {
+  return const_iterator(m_head ? this->m_head : this->m_end);
+}
+
+template <class T>
+typename s21::list<T>::iterator s21::list<T>::end() noexcept {
+  return iterator(m_end);
+}
+
+template <typename T>
+typename s21::list<T>::const_iterator s21::list<T>::cend() const noexcept {
+  return const_iterator(m_end);
+}
+
+template <typename T>
+s21::list<T>::list() noexcept : m_head(nullptr), m_tail(nullptr), m_size(0) {
+  m_end = new Node(T());
+  update_end();
+}
+
+template <typename T>
+s21::list<T>::list(size_type n) : m_head(nullptr), m_tail(nullptr), m_size(n) {
+  if (n <= 0) throw std::out_of_range("Size must be greater than zero");
```

Comment:
1. Это условие никогда не будет истинным. Вообще. Т.к. `size_t` - это беззнаковый тип и его значения не могут быть отрицательными. Даже если ты передашь как аргумент значение `-10`. В битовом плане оно всё равно будет трактоваться как положительное число и превратиться в `1323656467` какое-нибудь.
2. Тут лучше подойдёт тип исключения `std::length_error`.
3. Нет проверки на верхнюю границу допустимого значения. Условно говоря, что будет если укажут размер в 100000000? Хватит ли памяти? Вот в `size_t` такое число поместится. Но должно ли оно быть допустимым для создания списка? Мне кажется, что нужно установить какие-то границы. Которые желательно определять динамически в зависимости от платформы. Хотя, в любом случае, `new` выкинет исключение если памяти не хватит. Так что это максимально опциональная фигня. Наверное, лучше ограничиться исключением в `new`. Но если начали думать об области допустимых значений для размера, то можно было бы и о верхней границе подумать.

```suggestion
  if (n <= 0) throw std::out_of_range("Size must be greater than zero");
```